File: ios/Packages/FeatureServices/DeviceService/GemstoneDeviceStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.Device
import protocol Gemstone.GemDeviceStore
import GemstonePrimitives
import Preferences
import Primitives

public final class GemstoneDeviceStore: GemDeviceStore, @unchecked Sendable {
    private let preferences: Preferences

    public init(preferences: Preferences = .standard) {
        self.preferences = preferences
    }

    public func isRegistered() async throws -> Bool {
        preferences.isDeviceRegistered
    }

    public func setRegistered(registered: Bool) async throws {
        preferences.isDeviceRegistered = registered
    }

    public func getSubscriptionsVersion() async throws -> Int32 {
        preferences.subscriptionsVersion.asInt32
    }

    public func setSubscriptionsVersion(version: Int32) async throws {
        preferences.subscriptionsVersion = Int(version)
    }

    public func getPushedDevice() async throws -> Gemstone.Device? {
        preferences.pushedDevice
    }

    public func setPushedDevice(device: Gemstone.Device) async throws {
        preferences.pushedDevice = device
    }

    public func getPushedSubscriptions() async throws -> String? {
        preferences.pushedSubscriptions
    }

    public func setPushedSubscriptions(signature: String) async throws {
        preferences.pushedSubscriptions = signature
    }
}

File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/device/GemstoneDeviceStore.kt
package com.gemwallet.android.data.repositories.device

import com.gemwallet.android.data.service.store.ConfigStore
import uniffi.gemstone.GemDeviceStore

class GemstoneDeviceStore(
    private val configStore: ConfigStore,
) : GemDeviceStore {

    override suspend fun isRegistered(): Boolean = configStore.getBoolean(KEY_REGISTERED)

    override suspend fun setRegistered(registered: Boolean) = configStore.putBoolean(KEY_REGISTERED, registered)

    override suspend fun getSubscriptionsVersion(): Int = configStore.getInt(KEY_SUBSCRIPTION_VERSION)

    override suspend fun setSubscriptionsVersion(version: Int) = configStore.putInt(KEY_SUBSCRIPTION_VERSION, version)

    override suspend fun getPushedDevice(): String? = configStore.getString(KEY_PUSHED_DEVICE).takeIf { it.isNotEmpty() }

    override suspend fun setPushedDevice(device: String) = configStore.putString(KEY_PUSHED_DEVICE, device)

    override suspend fun getPushedSubscriptions(): String? = configStore.getString(KEY_PUSHED_SUBSCRIPTIONS).takeIf { it.isNotEmpty() }

    override suspend fun setPushedSubscriptions(signature: String) = configStore.putString(KEY_PUSHED_SUBSCRIPTIONS, signature)

    private companion object {
        const val KEY_REGISTERED = "device_registered"
        const val KEY_SUBSCRIPTION_VERSION = "subscription_version"
        const val KEY_PUSHED_DEVICE = "pushed_device"
        const val KEY_PUSHED_SUBSCRIPTIONS = "pushed_subscriptions"
    }
}
